Initial feedback

kylebrowning opened the pull request.

hartbit commented on Sources/NIOAPNS/APNSConfiguration.swift:
```diff
@@ -0,0 +1,41 @@
+    public let sslContext: NIOOpenSSL.SSLContext
+
+    public var url: URL {
+        get {
```
Nitpick: the `get` block is superfluous

kylebrowning replied:
👍

hartbit commented on Sources/NIOAPNS/APNSErrors.swift:
```diff
@@ -11,35 +11,35 @@ public struct APNSError: Codable {
-    case ServiceUnavailable = "ServiceUnavailable"
-    case Shutdown = "Shutdown"
-    case Unknown = "unknown"
+    case badCollapseId = "BadCollapseId"
```
What do you think about un-abbreviating the case name (`badCollapseIdentifier`)?

kylebrowning replied:
No objections on the left side, but the value is required to as thats what is returned from Apple's servers.

hartbit commented on Sources/NIOAPNS/APNSRequest.swift:
```diff
@@ -9,16 +9,18 @@ import NIO
-    public let aps: Aps
-    public let custom: [String: String]?
-    public init(aps: Aps, custom: [String: String]?) {
+public protocol APNotification: Codable {
```
I'd go for `Notification` or `APNSNotification` personally. I know this is all kind of subjective.

kylebrowning replied:
@hartbit The only reason I didn't want to do that was because of the namespace.

What if there is another Class/Struct named Notification?

hartbit replied:
There's always a way for users to disambiguate and I don't think there is a big risk with `Notification`.

kylebrowning replied:
@hartbit For the record, going with that means, I need to now do.
`NIOAPNS.Notification`

```swift
struct AcmeNotification: NIOAPNS.Notification {
    let acme2: [String]
    let aps: APSPayload
    
    init(acme2: [String], aps: APSPayload) {
        self.acme2 = acme2
        self.aps = aps
    }
}
```

hartbit replied:
Why? What is the other `Notification` its clashing with?

kylebrowning replied:
Foundation.

https://imgur.com/a/zw8SZlT

hartbit replied:
Oops. Okay, we've got no choice then. 😞 What about `APNSNotification` then?

kylebrowning replied:
Sure.

hartbit commented on Sources/NIOAPNS/APNSRequest.swift:
```diff
@@ -9,16 +9,18 @@ import NIO
+    var aps: APSPayload { get }
+}
+
+public struct APNS: APNotification {
```
The above protocol could be named `NotificationProtocol` and this one `Notification` to echo the Standard Library's `String/StringProtocol`.

kylebrowning replied:
Ill add the protocol bit, but see above.

What if there is another Class/Struct named `Notification`?

hartbit commented on Sources/NIOAPNS/APNSRequestEncoder.swift:
```diff
@@ -10,33 +10,44 @@ import NIO
 import NIOHTTP2
 
+// Can someone explain why I could do this, but not
+// let wrappedRequest = try! JSONEncoder().encode(req)
```
`APNotification` is a protocol, and `JSONEncoder().encode` expects a generic type. You need to make your encoder generic over the notification:

```swift
internal final class APNSRequestEncoder<T: APNotification> : ChannelOutboundHandler {
    // ...
    func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
        let req: T = unwrapOutboundIn(data)
	let wrappedRequest = try! JSONEncoder().encode(req)
        // ...
    }
}
```

kylebrowning replied:
Thanks, fixed.

kylebrowning replied:
`Cannot convert value of type 'APNSRequestEncoder<T>.OutboundIn' (aka 'APNotification') to specified type 'T'`

kylebrowning replied:
So if I do `as! T` that solves that, but then, On the `send<T: APNSNotifiction>` I get 
```Value of type 'APNSRequestEncoder<_>' does not conform to 'ChannelHandler' in coercion```

on 

```APNSRequestEncoder(deviceToken: deviceToken, apnsConfig: self.configuration)```

hartbit replied:
OutboundIn needs to be typealiased to T

kylebrowning replied:
Cannot invoke 'encode' with an argument list of type '(APNSRequestEncoder<T>.T)'

hartbit replied:
Hmm, I got it working... See my commit: https://github.com/hartbit/swift-nio-http2-apns/commit/4f13fb8c42602ed4f22450a375d5a2adf268fc48

tanner0101 commented on Sources/NIOAPNS/APNSConfiguration.swift:
```diff
@@ -0,0 +1,41 @@
+    public let signingMode: SigningMode
+    public let topic: String
+    public let environment: APNSEnvironment
+    public let sslContext: NIOOpenSSL.SSLContext
```
I would store `TLSConfiguration` here instead, then create the `SSLContext` later once you need it. That way you can initialize multiple clients with one copy of `APNSConfiguration`

tanner0101 commented on Sources/NIOAPNS/APNSConnection.swift:
```diff
@@ -3,14 +3,14 @@ import NIOHTTP2
 
 final public class APNSConnection {
-    public static func connect(apnsConfig: APNSConfig, on eventLoop: EventLoop) -> EventLoopFuture<APNSConnection> {
+    public static func connect(configuration: APNSConfiguration, on eventLoop: EventLoop) -> EventLoopFuture<APNSConnection> {
```
`configuration` is kind of verbose here, but I agree with previous comments that we should avoid abbreviations. Is there precedent for doing something like `using:` or `with:` or just `_:`? Not a big deal either way.

hartbit replied:
I know it's kind of verbose, but I worry that it wouldn't be as clear with `using:` or `with:`. But I agree, it's not a big deal.

tanner0101 commented on Sources/NIOAPNS/APNSConnection.swift:
```diff
@@ -19,27 +19,27 @@ final public class APNSConnection {
     }
     
-    public func send(deviceToken: String, _ request: APNSRequest) -> EventLoopFuture<APNSResponse> {
+    public func send<T: APNotification>( _ request: T, deviceToken: String) -> EventLoopFuture<APNSResponse> {
```
I prefer putting the generic constraints after the method, i.e.:
```swift
public func send<T>(...) where T: APNotification { ... }
```

tanner0101 replied:
Also, Swift style tends to prefer readable names for generic args, not `T` like in Java (although Vapor is definitely guilty of this). This would probably be better as `Notification`.

hartbit replied:
Yeah, I think we need `Notification` in this case.

hartbit replied:
Concerning the generic constraint, the Standard Library tends to keep simple constraints before the arguments, so this doesn't bother me personally.

tanner0101 commented on Sources/NIOAPNS/APNSRequestEncoder.swift:
```diff
@@ -10,33 +10,44 @@ import NIO
+// Cannot invoke 'encode' with an argument list of type '(APNotification)'
+
+extension Encodable {
+    func toJSONData() -> Data? {
```
This extension is only used once, I don't think it's necessary.

tanner0101 commented on Sources/NIOAPNS/APNSRequestEncoder.swift:
```diff
@@ -10,33 +10,44 @@ import NIO
-        let req:APNSRequest = unwrapOutboundIn(data)
-        let wrappedRequest = try! JSONEncoder().encode(req)
+        let req:APNotification = unwrapOutboundIn(data)
+        let wrappedRequest = req.toJSONData()!
```
It would be better to capture the error here and fail the write promise with the error so the user can see why it failed.

hartbit replied:
Agreed, especially as we don't control the type being encoded.

tanner0101 commented on Sources/NIOAPNS/APNSRequestEncoder.swift:
```diff
@@ -10,33 +10,44 @@ import NIO
     typealias OutboundOut = HTTPClientRequestPart
     
-    public let apnsConfig: APNSConfig
+    public let apnsConfig: APNSConfiguration
```
This should probably be renamed to `configuration` as well.

hartbit replied:
I'm not too bothered with `APNSConfiguration` here. `Configuration` will definitely be a common identifier and there's precedent with `URLSessionConfiguration`. It's good to know what its a configuration of.

kylebrowning replied:
I think maybe he's referring to the property name?

hartbit replied:
In that case, I agree :)

kylebrowning commented:
@hartbit @tanner0101 Okay I've updated this PR with all the comments and feedback. LMK.

I also addressed all the force unwraps and convert them to throw appropriately, and write to the promise where necessary.

hartbit commented on Sources/NIOAPNS/APNSConnection.swift:
```diff
@@ -19,27 +20,27 @@ final public class APNSConnection {
     }
     
-    public func send(deviceToken: String, _ request: APNSRequest) -> EventLoopFuture<APNSResponse> {
+    public func send<T: APNSNotificationProtocol>( _ request: T, deviceToken: String) -> EventLoopFuture<APNSResponse> {
```
The first parameter should probably be called `notification` now. Also, what do you think of renaming the label for the second one to `to`. I think it reads nicely:

```
connection.send(notification, to: userDeviceToken)
```

kylebrowning replied:
oh Good call, I missed that comment. Thanks!

hartbit commented on Sources/NIOAPNS/APNSRequestEncoder.swift:
```diff
@@ -10,34 +10,47 @@ import NIO
+        do {
+            data = try JSONEncoder().encode(req)
+        } catch {
+            promise?.fail(error: APNSCoderError.encodingFailed)
```
How about failing with the encoding error directly? I don't think we gain much by wrapping it: we actually loose the information on why it failed.

kylebrowning replied:
Good call. :)

hartbit commented on Sources/NIOAPNS/DataSigner.swift:
```diff
@@ -11,7 +11,7 @@ import CAPNSOpenSSL
     private let opaqueKey: OpaquePointer
 
-    public init?(data: Data) {
+    public init (data: Data) throws {
```
nitpick: I think the extra space here was a mistake?

hartbit commented on Sources/NIOAPNS/FileSigner.swift:
```diff
@@ -8,8 +8,13 @@
+    public convenience init (url: URL) throws {
+        let data: Data
+        do {
+            data = try Data.init(contentsOf: url)
```
Very minor nitpick: you could `try self.init(data: Data(contentsOf: url))` directly in the do block

kylebrowning replied:
Is the do block even needed?

```
public class FileSigner: DataSigner {
    public convenience init(url: URL) throws {
        try self.init(data: try Data.init(contentsOf: url))
    }
}
```

hartbit replied:
It's needed if you want to return your custom error on failure instead of the `Data` initialiser error.

hartbit commented on Sources/NIOAPNS/SigningMode.swift:
```diff
@@ -15,11 +15,11 @@ public struct SigningMode {
 extension SigningMode {
-    public static func file(path: String) -> SigningMode {
-        return .init(signer: FileSigner(url: URL(fileURLWithPath: path))!)
+    public static func file(path: String) throws  -> SigningMode {
```
Super minor nitpick: double space here probably unintentional

hartbit commented:
@kylebrowning Great work! Thanks for being so welcoming with the feedback 🙇‍♂️

tanner0101 commented on Sources/NIOAPNS/APNSConnection.swift:
```diff
@@ -3,14 +3,15 @@ import NIOHTTP2
             .channelOption(ChannelOptions.socket(IPPROTO_TCP, TCP_NODELAY), value: 1)
             .channelInitializer { channel in
-                let sslHandler = try! OpenSSLClientHandler(context: apnsConfig.sslContext, serverHostname: apnsConfig.getUrl().host)
+                let sslContext = try! SSLContext(configuration: configuration.tlsConfiguration)
```
I think the best way to handle these ones is doing something like:
```swift
let sslHandler: OpenSSLClientHandler
do {
    let sslContext = try SSLContext(configuration: configuration.tlsConfiguration)
    sslHandler = try OpenSSLClientHandler(context: sslContext, serverHostname: configuration.url.host)
} catch {
    // TODO: replace print statement w/ apple/logging
    print("Could not create TLS handler: \(error)")
    channel.close(mode: .all, promise: nil)
    return 
}
```

kylebrowning replied:
@tanner0101  have to return an event loop future here.

tanner0101 replied:
ah yeah, even better then, return `channel.eventLoop.makeFailedFuture(tlsError)`

kylebrowning replied:
Okay Cool!

tanner0101 commented on Sources/NIOAPNS/APNSRequestEncoder.swift:
```diff
@@ -10,34 +10,47 @@ import NIO
+            data = try JSONEncoder().encode(req)
+        } catch {
+            promise?.fail(error: APNSCoderError.encodingFailed)
+            ctx.close(promise: promise)
```
You should pass `nil` here since the promise was already completed in the previous line. (Promises can only be completed once)

tanner0101 commented on Sources/NIOAPNS/APNSRequestEncoder.swift:
```diff
@@ -10,34 +10,47 @@ import NIO
+            token = try jwt.sign(with: configuration.signingMode)
+        } catch {
+            promise?.fail(error: APNSTokenError.tokenWasNotGeneratedCorrectly)
+            ctx.close(promise: promise)
```
same here regarding passing the promise

tanner0101 commented on Sources/NIOAPNSExample/main.swift:
```diff
@@ -5,25 +5,37 @@ import NIOOpenSSL
-                            env: .sandbox)
+let apnsConfig = APNSConfiguration(keyIdentifier: "2M7SG2MR8K",
+                                   teamIdentifier: "ABBM6U9RM5",
+                                   signingMode: try .file(path: "/Users/kylebrowning/Downloads/key.p8"),
```
I tend to prefer the style of putting the `try` all the way to the left, but maybe that's just me. For example:
```swift
let apnsConfig = try APNSConfiguration(...)
```

kylebrowning commented:
@tanner0101 @hartbit moar fixes!

kylebrowning commented:
Didn't mean to close sorry.

tanner0101 commented on Sources/NIOAPNS/APNSConnection.swift:
```diff
@@ -3,43 +3,50 @@ import NIOHTTP2
-                    multiplexer
-                ]
-                return channel.pipeline.addHandlers(handlers, first: false)
+                let sslHandler: OpenSSLClientHandler
```
this doesn't need to be declared outside anymore

kylebrowning replied:
Great catch.

kylebrowning commented:
merged, thanks!
